website/addons/owncloud/model.py

```diff
@@ -0,0 +1,153 @@
+# -*- coding: utf-8 -*-
+import logging
+from modularodm import fields
+from framework.auth import Auth
+from website.addons.base import exceptions
+from website.addons.base import StorageAddonBase
+from website.addons.base import (
+    AddonOAuthNodeSettingsBase, AddonOAuthUserSettingsBase,
+)
+from website.addons.owncloud.serializer import OwnCloudSerializer
+from website.addons.owncloud.settings import DEFAULT_HOSTS, USE_SSL
+from owncloud import Client as OwnCloudClient
+from website.addons.owncloud import settings
+
+from website.oauth.models import BasicAuthProviderMixin
+from website.util import api_v2_url
+
+logger = logging.getLogger(__name__)
+
+class OwnCloudProvider(BasicAuthProviderMixin):
+    """An alternative to `ExternalProvider` not tied to OAuth"""
+
+    name = 'ownCloud'
+    short_name = 'owncloud'
+    serializer = OwnCloudSerializer
+
+    def __repr__(self):
+        return '<{name}: {status}>'.format(
+            name=self.__class__.__name__,
+            status=self.account.display_name if self.account else 'anonymous'
+        )
+
+class AddonOwnCloudUserSettings(AddonOAuthUserSettingsBase):
+    oauth_provider = OwnCloudProvider
+    serializer = OwnCloudSerializer
+
+    def to_json(self, user):
+        ret = super(AddonOwnCloudUserSettings, self).to_json(user)
+        ret['hosts'] = DEFAULT_HOSTS
+        return ret
+
+class AddonOwnCloudNodeSettings(StorageAddonBase, AddonOAuthNodeSettingsBase):
+    oauth_provider = OwnCloudProvider
+    serializer = OwnCloudSerializer
+
+    folder_id = fields.StringField()
+
+    _api = None
+
+    @property
+    def api(self):
+        """authenticated ExternalProvider instance"""
+        if self._api is None:
+            self._api = OwnCloudProvider(self.external_account)
+        return self._api
+
+    @property
+    def folder_path(self):
+        return self.folder_id
+
+    @property
+    def folder_name(self):
+        return self.fetch_folder_name()
+
+    def set_folder(self, folder, auth=None):
+        if folder == '/ (Full ownCloud)':
+            folder = '/'
+        self.folder_id = folder
+        self.save()
+        if auth:
+            self.nodelogger.log(action='folder_selected', save=True)
+
+    def fetch_folder_name(self):
+        if self.folder_id == '/':
+            return '/ (Full ownCloud)'
+        return self.folder_id
+
+    def clear_settings(self):
+        self.folder_id = None
+
+    def deauthorize(self, auth=None, add_log=True):
+        """Remove user authorization from this node and log the event."""
+        self.clear_settings()
+        self.clear_auth()
+        if add_log and auth:
```

Comment:
>So, the difference between the logs and the anonymous logs is done is the js layer?

Erm, not exactly. Anonymous logs are only shown when an "Anonymous View-Only Link` is used to authenticate and view a page (these are project-specific and generated on the Contributors page). They're intended for things like getting peer review anonymously, to prevent as much bias as possible.

It is, however, possible for a log that includes `${user}` in its JSON definition to be generated -- I think if another contributor removes the addon-authorizing contributor, one of these logs will be generated without `auth`, which would either cause `LogText.view.logText` to append `_no_user` to the log's `.action`, or cause `LogPieces.user` to evaluate to `A user` rather than the user's name (I don't recall exactly how it works).

```python
def deauthorize(self, auth=None, add_log=True):
    self.clear_settings()
    self.nodelogger.log(action='node_deauthorized')
    self.clear_auth()  # Also performs a .save()
```